### PR TITLE
Visibility Simulator Overhaul

### LINKED ISSUE #154: Make vis_cpu optional
Most of the core functionality doesn't require it.

### PR DESCRIPTION
**Pull request by steven-murray**
This PR should do the job of the previous #162 and #163, and hopefully not change EVERY file.

Fixes #154

* It fixes `VisCPU` and `HealVis` to use the new setup
* Adds `UVSim` simulator.

@hughbg can you check especially if the call to correct_point_source_pos is good? And also if polarization is being used correctly?

### CODE REVIEW

File: hera_sim/visibilities/vis_cpu.py [hunk 1]
```diff
@@ -315,38 +227,48 @@ def _base_simulate(self, crd_eq, I_sky):
...
         # Get pixelized beams if required
         if self.use_pixel_beams:
-            beam_lm = self.get_beam_lm()
+            beam_lm = self.get_beam_lm(data_model)
         else:
             beam_list = [
-                self.beams[np.where(self.beam_ids == ant)[0][0]]
-                for ant in self.ant_list
+                data_model.beams[np.where(data_model.beam_ids == ant)[0][0]]
+                for ant in data_model.ant_list
```
Comment (r-pascua) [inline hera_sim/visibilities/vis_cpu.py:249]: This isn't going to work with the way the `beam_ids` array is constructed, since `beam_ids` won't necessarily have a unique entry for each antenna.

I think I'm a little too tired (and my head too stuck in the mist of edge cases) to figure out the right way to do this right now, but the thing I'm worrying about is the case where the antenna numbers don't monotonically increase in the `UVData` object. I'm mainly concerned about ensuring that we get the ordering right on all the moving parts here, so that the beams get matched with the right antennas and the visibilities returned by the simulator match the blt-ordering imposed on the data array.

**Review by r-pascua — COMMENTED** [review]
This is looking really great! A huge improvement from where it was before. I made a handful of comments, some that I think are important to get to, others not so much.

I'm running out of mental steam right now, but I definitely think it would be worthwhile for us to sit down and brainstorm the right way to handle how various arrays are sorted, because getting that wrong leads to nasty silent bugs.

Anyway, thank you for your work! I really like the direction this is going.

File: hera_sim/visibilities/simulators.py [hunk 2]
```diff
@@ -13,374 +12,129 @@
...
-        beams=None,
-        beam_ids=None,
-        sky_intensity=None,
-        point_source_pos=None,
-        point_source_flux=None,
-        nside=2 ** 5,
-        validate=True,
+        *,
+        uvdata: UVData,
+        sky_model: SkyModel,
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:63]: The idea is to make sure they're always passed as keywords. This is good for future-proofing the API (in case we change parameter ordering in any way).

File: hera_sim/visibilities/simulators.py [hunk 3]
```diff
@@ -13,374 +12,129 @@
...
+        beam_ids: Dict[str, int] | None = None,
+        beams: BeamListType | None = None,
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:65]: Apparently this is the preferred syntax in the annotations you import from `__future__`.

File: hera_sim/visibilities/simulators.py [hunk 4]
```diff
@@ -13,374 +12,129 @@
...
-
-            self.beams = [ab.AnalyticBeam("uniform")] if beams is None else beams
-            if beam_ids is None:
+        self.uvdata = uvdata
+        self.beams = [ab.AnalyticBeam("uniform")] if beams is None else beams
+        self.n_ant = (
+            self.uvdata.Nants_data
+        )  # NOT Nants because we only want ants with data
+
+        assert isinstance(self.uvdata, UVData)
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:74]: Fair enough. Done in aca8de5

File: hera_sim/visibilities/simulators.py [hunk 5]
```diff
@@ -13,374 +12,129 @@
...
-        """
-        hmap = np.zeros((len(point_source_flux), aph.nside_to_npix(nside)))
-
-        # Get which pixel every point source lies in.
-        pix = aph.lonlat_to_healpix(
-            lon=point_source_pos[:, 0] * units.rad,
-            lat=point_source_pos[:, 1] * units.rad,
-            nside=nside,
+        else:
+            self.beam_ids = np.array(beam_ids, dtype=int)
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:88]: I was actually considering whether we instead preferred to keep the beam_ids dict from pyuvsim instead of the list that we turn it into. Or perhaps keep both around on the instance. We can allow both as input at least, good point.

The `ModelData` class is going to be used by the user to build up a `VisibilitySimulation`.

[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py]: I've updated accept the dict input. Will await your ideas on whether we should keep the dict format around (or switch to it).

File: hera_sim/visibilities/simulators.py [hunk 6]
```diff
@@ -13,374 +12,129 @@
...
+
+        assert isinstance(self.beam_ids, np.ndarray)
+        assert self.beam_ids.dtype == int
+        assert self.beam_ids.max() < self.n_beams
+        assert len(self.beam_ids) == self.n_ant
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py]: I agree. I've now done this.

File: hera_sim/visibilities/simulators.py [hunk 7]
```diff
@@ -13,374 +12,129 @@
...
+
+        self.sky_model = sky_model
+        self.sky_model.at_frequencies(self.freqs * units.Hz)
+        assert isinstance(self.sky_model, SkyModel)
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py]: Good catch. Fixed.

File: hera_sim/visibilities/simulators.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py]: Another thing to think about: in the the `from_config` method, we do this to get the `beam_ids` from the dict:

```python
nms = list(uvdata.antenna_names)
tmp_ids = np.zeros(len(beam_ids), dtype=int)
for name, beam_id in beam_ids.items():
    tmp_ids[nms.index(name)] = beam_id
beam_ids=tmp_ids
```

Instead of simply `np.array(list(beam_ids.values()), dtype=int)`. I wonder if we did this for a reason -- perhaps we were ensuring ordering for pre-py36, where dicts weren't ensured to have a particular ordering?

[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py]: Just to be sure, I changed the conversion in `_process_beams` to take account of this ordering, but simplified the above into this:

```python
beam_ids = np.array(
    [beam_ids[nm] for nm in self.uvdata.antenna_names], dtype=int
)
```

File: hera_sim/visibilities/simulators.py [hunk 8]
```diff
@@ -415,3 +169,112 @@ def write_config_file(
             return_names=False,
             path_out=direc,
         )
+
+    @cached_property
+    def ant_list(self) -> np.ndarray:
+        """An orderd list of active antenna numbers."""
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:208]: Done, thanks.

File: hera_sim/visibilities/simulators.py [hunk 9]
```diff
@@ -415,3 +169,112 @@ def write_config_file(
...
+class VisibilitySimulator(metaclass=ABCMeta):
+    """Base class for all hera_sim compatible visibility simulators."""
+
+    #: Whether this particular simulator has the ability to simulate point
+    #: sources directly.
+    point_source_ability = True
+
+    #: Whether this particular simulator has the ability to simulate diffuse
+    #: maps directly.
+    diffuse_ability = False
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:304]: Yeah, that's true. Then they just have to set this attribute properly. I think most simulators assume input point sources by default.

File: hera_sim/visibilities/simulators.py [hunk 10]
```diff
@@ -415,3 +169,112 @@ def write_config_file(
...
+
+    @abstractmethod
+    def simulate(self, data_model: ModelData) -> np.ndarray:
+        """Simulate the visibilities."""
+        pass
+
+    def validate(self, data_model: ModelData):
+        """Check that the data model complies with the assumptions of the simulator."""
+        pass
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:313]: Yeah the default `validate()` is to do nothing. Many simulators might not have to do any validation (i.e. they might allow any kind of data as input).

File: hera_sim/visibilities/healvis_wrapper.py [hunk 11]
```diff
@@ -36,45 +37,45 @@ class HealVis(VisibilitySimulator):
...
-        if "beams" not in kwargs:
-            kwargs["beams"] = [AnalyticBeam("uniform")]
+    def validate(self, model_data: ModelData):
+        """Validate that all data is correct.
 
-        super(HealVis, self).__init__(**kwargs)
+        In addition to standard parameter restrictions, HealVis requires a single beam
+        for all antennae.
+        """
+        assert model_data.n_beams == 1
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/healvis_wrapper.py:56]: Done.

File: hera_sim/visibilities/healvis_wrapper.py [hunk 12]
```diff
@@ -36,45 +37,45 @@ class HealVis(VisibilitySimulator):
...
+        if isinstance(model_data.beams[0], pyuvsim.analyticbeam.AnalyticBeam):
+            old_args = model_data.beams[0].__dict__
 
             gauss_width = None
             if old_args["type"] == "gaussian":
                 if old_args["sigma"] is None:
                     raise NotImplementedError(
-                        "Healvis does not permit " "gaussian beam with diameter."
+                        "Healvis does not permit gaussian beam with diameter."
                     )
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/healvis_wrapper.py:68]: Yeah, I think this was something that happened in two stages. I've simplified it now.

File: hera_sim/visibilities/healvis_wrapper.py [hunk 13]
```diff
@@ -103,25 +94,30 @@ def sky_model(self):
...
+            intensity = 10 ** -26 * sky_model.stokes[0].T.value
+            intensity *= (
+                cnst.c.to("m/s").value / sky_model.freq_array.to("Hz").value
+            ) ** 2 / (2 * cnst.k_B.value)
+        elif sky_model.stokes.unit == units.K:
+            intensity = sky_model.stokes[0].T.value
+        else:
+            raise ValueError(
+                f"Units of {sky_model.stokes.unit} are not compatible with healvis"
+            )
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/healvis_wrapper.py:114]: I like that. Done.

File: hera_sim/visibilities/healvis_wrapper.py [hunk 14]
```diff
@@ -146,15 +142,12 @@ def _simulate(self):
...
-
+        obs = self.get_observatory(data_model)
+        sky = self.get_sky_model(data_model.sky_model)
+        visibility = [
+            obs.make_visibilities(sky, Nprocs=self._nprocs, beam_pol=pol)[0]
+            for pol in data_model.uvdata.get_pols()
+        ]
         visibility = np.moveaxis(visibility, 0, -1)
 
         return visibility[:, 0][:, np.newaxis, :, :]
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/healvis_wrapper.py:153]: OK, yeah I see it. This will have to be fixed before merging.

File: hera_sim/visibilities/simulators.py [hunk 15]
```diff
@@ -1,10 +1,9 @@
 """Module defining a high-level visibility simulator wrapper."""
-from __future__ import division
-import warnings
+from __future__ import division, annotations
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:2]: The division shouldn't be there -- removed it now. The `annotations` are important for some nice features that are coming in py3.10+ (like using a class as a type within its own body, and using `|` for `Union`).

File: hera_sim/visibilities/simulators.py [hunk 16]
```diff
@@ -415,3 +169,112 @@ def write_config_file(
...
+            stokes=np.array(
+                [
+                    hmap.value,
+                    np.zeros_like(hmap),
+                    np.zeros_like(hmap),
+                    np.zeros_like(hmap),
+                ]
+            )
+            * units.Jy
+            / units.sr,
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:266]: Yeah. I tried to change it back but it wouldn't have it. 🤷🏻

File: hera_sim/visibilities/simulators.py [hunk 17]
```diff
@@ -415,3 +169,112 @@ def write_config_file(
...
+            hpx_inds=np.arange(npix),
+            spectral_type="full",
+            freq_array=sky_model.freq_array,
+        )
+
+    def _write_history(self):
+        """Write pertinent details of simulation to the UVData's history."""
+        class_name = self.simulator.__class__.__name__
+        self.uvdata.history += (
+            f"Visibility Simulation performed with hera_sim's {class_name} simulator\n"
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:278]: Yes.

File: hera_sim/visibilities/simulators.py [hunk 18]
```diff
@@ -415,3 +169,112 @@ def write_config_file(
...
+        )
+        self.uvdata.history += f"Class Repr: {repr(self.simulator)}"
+
+    def simulate(self):
+        """Perform the visibility simulation."""
+        self._write_history()
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/simulators.py:284]: Hmmmm good point. I'll change it to afterwards.

File: hera_sim/visibilities/vis_cpu.py [hunk 19]
```diff
@@ -48,18 +48,21 @@ class VisCPU(VisibilitySimulator):
...
+        precision: int = 1,
+        use_gpu: bool = False,
         mpi_comm=None,
-        **kwargs
+        ref_time: Optional[Union[str, Time]] = None,
+        correct_source_positions: bool = False,
     ):
 
-        assert precision in (1, 2)
+        assert precision in {1, 2}
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py:65]: Hmmm, that'll have to be coordinated with the `vis_cpu` package.

File: hera_sim/visibilities/vis_cpu.py [hunk 20]
```diff
@@ -92,101 +95,40 @@ def __init__(
...
         """Checks for correct input format."""
-        super(VisCPU, self).validate()
-
-        # This one in particular requires that every baseline is used!
-        N = len(self.uvdata.get_ants())
-
         # N(N-1)/2 unique cross-correlations + N autocorrelations.
-        if len(self.uvdata.get_antpairs()) != N * (N + 1) / 2:
+        if (
+            len(data_model.uvdata.get_antpairs())
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py:105]: Are you sure this is guaranteed to be the same? I don't have enough experience with pyuvdata, but I know there's sometimes differences between active/non-active antennas etc.

File: hera_sim/visibilities/vis_cpu.py [hunk 21]
```diff
@@ -92,101 +95,40 @@ def __init__(
...
                 "VisCPU requires using every pair of antennas, "
                 "but the UVData object does not comply."
             )
 
-        if len(self.uvdata.data_array) != len(self.uvdata.get_antpairs()) * len(
-            self.lsts
-        ):
+        if len(data_model.uvdata.data_array) != len(
+            data_model.uvdata.get_antpairs()
+        ) * len(data_model.lsts):
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py:115]: I think so. That's why they have a blt axis instead of splitting that into two axes.

File: hera_sim/visibilities/vis_cpu.py (hunk 1 above)
Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py:249]: Yup. I'm with you. I'm wary of all of this, because I know there's a few moving parts. Perhaps we should chat live about it, and maybe rope in @mkolopanis  / @bhazelton .

**Comment by steven-murray** [comment]
Thanks @r-pascua for an amazing review. I've addressed all the easy stuff, so feel free to go through and resolve those things if they seem adequate. The stuff that's left is mostly the beam_id ordering, which I think we should chat live about.

File: hera_sim/visibilities/simulators.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/simulators.py]: Ah, gotcha! I think this is a very good move, then.

File: hera_sim/visibilities/simulators.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/simulators.py]: Huh, ok

File: hera_sim/visibilities/simulators.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/simulators.py]: Nice, this is much more compact. I'm not sure why it was done the way it was before, but this looks like it achieves the same thing.

As for whether we should use the dict format from `pyuvsim` or not... I think I need to have a better picture of how everything works before weighing in on that.

File: hera_sim/visibilities/simulators.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/simulators.py:319]: Cool, thanks for the clarification!

File: hera_sim/visibilities/simulators.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/simulators.py]: Ah, interesting, ok! Thanks for clearing that up!

File: hera_sim/visibilities/simulators.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/simulators.py:271]: Well, if the robots say that's the way to do it...

File: hera_sim/visibilities/vis_cpu.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/vis_cpu.py:65]: Ah, ok.

File: hera_sim/visibilities/vis_cpu.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/vis_cpu.py:115]: Fair enough. Do you know if we can get the same check by comparing, say, `Nblts` and `Nbls * Ntimes`? I'm not familiar enough with using `UVData` objects set up like that to know...

File: hera_sim/visibilities/vis_cpu.py (hunk 1 above)
Comment (r-pascua) [reply hera_sim/visibilities/vis_cpu.py:249]: I think a live chat would be good. Maybe after next week's Validation meeting?

Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py:249]: Sounds good.

File: hera_sim/visibilities/vis_cpu.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (r-pascua) [reply hera_sim/visibilities/vis_cpu.py:105]: I haven't extensively tested it, nor have I looked at the code for how this attribute is retrieved, so I can't say for sure. If we're being cautious, then I think we're better off leaving as-is. (All I've done is a shallow check where I select only a few antennas and confirm that the `Nbls` attribute matches the length of the returned antenna pairs.)

**Comment by r-pascua** [comment]
OK, I've resolved all the comments that I think have been addressed. By my account, it looks like these are the remaining issues:
* Update the `lsts` attribute to use a more robust method of retrieving the unique LSTs (the current implementation only works for one particular blt-order).
* Do some work to ensure that the simulated visibilities from `healvis` and `vis_cpu` are reshaped so that they match the shape of the underlying `data_array`. I would like to write a unit test for this, but I'm not sure about the best way to implement this. One thought that comes to mind is checking that redundant baselines give identical visibilities (up to some tolerance) when identical beams are used, and that all of the autocorrelations match.
* Make sure that the `write` method notes the version of `hera_sim` that was used (and ideally also the version(s) of the simulator(s) used).
* Figure out how to ensure that beams, antenna numbers, and so-on are all sorted in the same way. We should do this in a live collaborative session.

Thanks again for the great work! My hope is that we resolve most (if not all) of these issues during the live discussion. Once they're fixed, I'll give the green light for merging this.

File: hera_sim/visibilities/vis_cpu.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py:105]: OK, looking at pyuvdata source code, it looks like `Nbls` is always correct. I'll change the code.

[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py]: (Done)

**Comment by r-pascua** [comment]
OK, if my accounting is correct, then we're very close to merging this. Here's what I think the remaining TODOs are:
* Update the `write` method to include the `hera_sim` version. (I feel like I'm missing something basic since I can't figure out how to do this in a non-circular way.)
* Update the beam handling to allow for `beam_ids` to be provided as an `array_like` object, subject to the constraint that it is either length 1 or has the same length as the number of active antennas (in which case we assume the IDs are sorted properly, and maybe issue a warning). This should fix the failing test.
* We still haven't gotten a review from @hughbg on whether the point source correction and polarization handling is correct.

Following this we have a few action items to complete before merging the overhaul into `main`. Here's my account of what we still need to complete after this PR is merged:
* Merge #170 
* Ensure that the new polarization features from #172 work with the refactor.
* Flesh out the unit tests to make sure we test that things work, and try to identify corner cases.

**Comment by steven-murray** [comment]
OK @r-pascua:

1. The visibility simulator already writes the version of hera_sim to the history. I've added code now to write a nominal version of the simulator itself.
2. I've added logic for beam_ids
3. I think we can deal with polarization handling when we merge up the latest `main`. 

We should actually merge this now -- it's not going to `main` yet.
